proposals/0000-unlifted-datatypes.rst

```diff
@@ -0,0 +1,482 @@
+check:
+
+::
+
+           if (R1 & 7 != 0) goto c1fi; else goto c1fj;
+       c1fj: // global
+           call (I64[R1])(R1) returns to c1fi, args: 8, res: 8, upd: 8;
+       c1fi: // global
+
+What does it do? It tests whether the scrutinee pointer in R1 has its tag bits
+(the lower 3 bits) set. If not, the code will *enter* the heap object to
+possibly evaluate it and get to know the constructor tag (i.e. 1 for ``Branch``
+and 2 for ``Leaf``). Now, most of the time, the heap object has already been
+evaluated and the pointer is in fact properly tagged (the GC for example will
+tag every pointer to an evaluated heap object).
+
+Nevertheless, we end up with the code above (which we loosely refer to as the
+"zero tag check", as in
+`#16820 <https://gitlab.haskell.org/ghc/ghc/issues/16820>`_), because a call
+site like ``member k undefined`` is possible! Obviously, ``undefined`` is
+neither evaluated nor tagged. Even making ``IntSet`` spine-strict by adding
+bangs to the ``Branch`` constructor doesn't get rid of the problem: ``member k
+undefined`` is still well-typed, so we have to generate code for it.
+Fundamentally, ⊥ is an inhabitant of every data type in Haskell, so every data
+declaration introduces a *lifted* type of kind ``TYPE (BoxedRep Lifted)``.
+
+So omission of the zero tag check is only possible if ⊥ is not an inhabitant of
+the type. This is exactly the condition for a type to be *unlifted*! Since we
+are still talking about heap objects, we need unlifted, but boxed types, which
+have kind ``TYPE (BoxedRep Unlifted)``. Indeed, for types of this kind GHC will
+already omit the zero tag check today! So all that is left is to extend GHC in
+a way that
+
+::
+  
+  data IntSet :: TYPE (BoxedRep Unlifted) where
+    Branch :: IntSet -> !Int -> IntSet -> IntSet
+    Leaf   :: IntSet
+
+  member :: Int -> IntSet -> Bool
+  member !k Leaf = False
+  member k (Branch l k' r)
+    | k == k'   = True
+    | k <  k'   = member k l
+    | otherwise = member k r
+
+does the Right Thing, namely disallowing unevaluated boxes to be passed around
+by using call-by-value. This way we can have the guarantee that all unlifted
+boxes are always properly tagged.
+
+Preliminary `benchmarks <https://gitlab.haskell.org/ghc/ghc/merge_requests/2218#note_239994>`_
+suggest that the unlifted variant is 5-10% faster, just due to omission of the
+zero tag check.
+
+Proposed Change Specification
+-----------------------------
+
+Henceforth, data type declaration refers to both data type and data family instance declarations.
+
+Static semantics
+~~~~~~~~~~~~~~~~
+
+Kind signatures for regular data type declarations must have a return kind of
+``TYPE (BoxedRep Lifted)``. Activating ``-XUnliftedDatatypes`` will lift this
+restriction to allow any return kind that unifies with ``TYPE (BoxedRep _)``. A
+data type explitly *may* have levity polymorphic kind
+``TYPE (BoxedRep l)`` for some type variable ``l``.
```

Comment:
This desperately needs an example:
```
data PList (l :: Levity) a :: TYPE (BoxedRep l) where
   PNil :: PList l a
   PCons :: a -> PList l a -> PList l a
```

Reply:
I thought I'd leave examples to the Examples section, where I give

```
  data List a :: TYPE (BoxedRep l) where
    Nil :: List a
    Cons :: a -> List a -> List a

  mapLifted :: (a -> b) -> List a -> List b
  mapLifted f Nil         = Nil
  mapLifted f (Cons x xs) = Cons (f x) (mapLifted f xs)
  
  mapUnlifted :: (a -> b) -> List @Unlifted a -> List @Unlifted b
  mapUnlifted f Nil         = Nil
  mapUnlifted f (Cons x xs) = Cons (f x) (mapLifted f xs)
 ```

I can give that example earlier, but my goal was to keep the Change Specification as concise as possible.

Reply:
I understand that goal, but feedback is that the Change Specification section is nigh incomprehensible without the examples.     The examples can be carefully chosen to illustrate the particular point being made by each sentence.   More distant examples just don't do that.

(In another world, you could imagine a change spec that included formal BNF and typing rules.  But even then examples are helpful.)